blueice/likelihood.py

```diff
@@ -101,29 +102,82 @@ def __init__(self, pdf_base_config, likelihood_config=None, **kwargs):
         # number of events per bin observed in Monte Carlo / calibration data that gave rise to the model.
         self.n_model_events_interpolator = lambda x: None
         self.n_model_events = None
 
+    @property
+    def source_shape_parameters(self):
+        """Dict of sources with shape parameters. source name -> dict of shape parameters."""
+        source_shape_parameters = OrderedDict()
+        for sn, source, apply_eff, eff_name in zip(
+                self.source_name_list,
+                self.base_model.sources,
+                self.source_apply_efficiency,
+                self.source_efficiency_names
+                ):
+            ignore_parameters = set(source.config['dont_hash_settings'])
+            # The efficiecny parameter doesn't need to be hashed but it needs to be passed to the morpher
+            if apply_eff:
+                ignore_parameters.discard(eff_name)
+            shape_parameters = OrderedDict({k: v for k, v in self.shape_parameters.items() if k not in ignore_parameters})
+            if shape_parameters:
+                source_shape_parameters[sn] = shape_parameters
+        return source_shape_parameters
+
+    def _get_shape_indices(self, source_name):
+        """Return the indices of the shape parameters used by the source."""
+        shape_keys = self.source_shape_parameters[source_name].keys()
+        return [i for i, k in enumerate(self.shape_parameters.keys()) if k in shape_keys]
+
+    def _get_model_anchor(self, anchor, source_name):
+        """Return the shape anchors of the full model, given the shape anchors of a signle source.
+        All values of shape parameters not used by this source will be set to None.
+        """
+        shape_indices = self._get_shape_indices(source_name)
+        model_anchor = [None] * len(self.shape_parameters)
+        for i, idx in enumerate(shape_indices):
+            model_anchor[idx] = anchor[i]
+        return tuple(model_anchor)
+
     def prepare(self, n_cores=1, ipp_client=None):
         """Prepares a likelihood function with shape parameters for use.
         This will compute the models for each shape parameter anchor value combination.
         """
         if len(self.shape_parameters):
-            self.morpher = MORPHERS[self.config['morpher']](self.config.get('morpher_config', {}),
-                                                            self.shape_parameters)
-            zs_list = self.morpher.get_anchor_points(bounds=self.get_bounds())
+            if self.source_wise_interpolation:
+                # Create morphers for each source individually
+                self.source_morphers = OrderedDict()
+                for sn, shape_parameters in self.source_shape_parameters.items():
+                    self.source_morphers[sn] = MORPHERS[self.config['morpher']](
+                        self.config.get('morpher_config', {}),
+                        shape_parameters
```

Comment:
almost certainly overkill, but for readability, perhaps we can make a tiny `minimorpher = lambda shape_parameters: MORPHERS[self.config['morpher']]( self.config.get('morpher_config', {}), shape_parameters)` ?

Reply:
Good idea -- done! (though I made a proper def to make flake8 happy 😊